lib/Parse/ParseDeclCXX.cpp

```diff
@@ -1947,6 +1966,24 @@ void Parser::ParseClassSpecifier(tok::TokenKind TagTokKind,
         // We're parsing a reference to a generic struct, so we need to parse
         // the type arguments before we can instantiate.
         TagOrTempResult = ParseRecordTypeApplication(Decl);
+      } else if (Decl->isItypeGeneric()) {
+        if (Actions.IsCheckedScope() || Tok.is(tok::less)) {
+          // For an itype generic in a checked scope, we require type arguments as well.
+          // If the scope is unchecked but the user provides arguments, we allow that too.
+          TagOrTempResult = ParseRecordTypeApplication(Decl);
```

Comment:
This is correct, but the programmer could get confused by the resulting error messages if they forget the type application.  The same problem could happen with previous clause for IsGeneric().   The message would be something like "expected '<'", which won't be meaningful.

I think you need to modify ParseRecordTypeApplication to issue a diagnostic if the initial '<' is missing.  The error message should say that we expected a list of type arguments.  (the error message for missing type arguments for generic functions is `err_expected_type_argument_list_for_generic_call`, defined in include\clang\Basic\DiagnosticSemaKinds.td).